src/backend/distributed/utils/colocation_utils.c

```diff
@@ -14,14 +14,322 @@
+	targetDistributionColumnType = targetDistributionColumn->vartype;
+
+	if (sourceDistributionColumnType != targetDistributionColumnType)
+	{
+		char *sourceRelationName = get_rel_name(sourceRelationId);
+		char *targetRelationName = get_rel_name(targetRelationId);
+
+		ereport(ERROR, (errmsg("distribution column types don't match for "
+							   "%s and %s", sourceRelationName, targetRelationName)));
+	}
+
+	/*
+	 * Get an exclusive lock on the colocation system catalog. Therefore, we
+	 * can be sure that there will no modifications on the colocation table
+	 * until this transaction is committed.
+	 */
+	pgDistColocation = heap_open(DistColocationRelationId(), ExclusiveLock);
+
+	/* check if shard placements are colocated */
+	ErrorIfShardPlacementsNotColocated(sourceRelationId, targetRelationId);
+
+	/*
+	 * Get colocation group of the source table, if the source table does not
+	 * have a colocation group, create a new one, and set it for the source table.
+	 */
+	sourceColocationId = TableColocationId(sourceRelationId);
+	if (sourceColocationId == INVALID_COLOCATION_ID)
+	{
+		uint32 shardCount = ShardIntervalCount(sourceRelationId);
+		uint32 shardReplicationFactor = TableShardReplicationFactor(sourceRelationId);
+
+		sourceColocationId = CreateColocationGroup(shardCount, shardReplicationFactor,
+												   sourceDistributionColumnType);
+		UpdateColocationGroup(sourceRelationId, sourceColocationId);
+	}
+
+	/* finally set colocation group for the target relation */
+	UpdateColocationGroup(targetRelationId, sourceColocationId);
+
+	heap_close(pgDistColocation, NoLock);
+}
+
+
+/*
+ * ErrorIfShardPlacementsNotColocated checks if the shard placements of the
+ * given two relations are physically colocated. If they are not colocated,
+ * it errors out.
+ *
+ * Note that, this functions assumes that both tables are hash distributed.
+ */
+static void
+ErrorIfShardPlacementsNotColocated(Oid leftRelationId, Oid rightRelationId)
+{
+	List *leftShardIntervalList = NIL;
+	List *rightShardIntervalList = NIL;
+	ListCell *leftShardIntervalCell = NULL;
+	ListCell *rightShardIntervalCell = NULL;
+	char *leftRelationName = NULL;
+	char *rightRelationName = NULL;
+	uint32 leftShardCount = 0;
+	uint32 rightShardCount = 0;
+
+	/* get sorted shard interval lists for both tables */
+	leftShardIntervalList = LoadShardIntervalList(leftRelationId);
+	rightShardIntervalList = LoadShardIntervalList(rightRelationId);
+
+	/* prevent concurrent placement changes */
+	LockShardListMetadata(leftShardIntervalList, ShareLock);
+	LockShardListMetadata(rightShardIntervalList, ShareLock);
+
+	leftRelationName = get_rel_name(leftRelationId);
+	rightRelationName = get_rel_name(rightRelationId);
+
+	leftShardCount = list_length(leftShardIntervalList);
+	rightShardCount = list_length(rightShardIntervalList);
+
+	if (leftShardCount != rightShardCount)
+	{
+		ereport(ERROR, (errmsg("shard counts don't match for %s and %s",
```

Comment:
Minor: Again, this looks more of an error detail. The actual error message should be `tables are not colocated` and we keep this message in the detail?